File: iOS/MeloRemote/Services/MusicAssistantClient.swift
import Foundation

actor MusicAssistantClient {
    typealias EventHandler = @Sendable (MAEventEnvelope) -> Void
    
    private let session: URLSession
    private var webSocket: URLSessionWebSocketTask?
    private var listenerTask: Task<Void, Never>?
    private var pingTask: Task<Void, Never>?
    private var pending: [
        String: CheckedContinuation<JSONValue, Error>
    ] = [:]
    
    private var timeoutTasks: [
        String: Task<Void, Never>
    ] = [:]
    
    private var partialResults: [String: [JSONValue]] = [:]
    private var eventHandler: EventHandler?
    private(set) var baseURL: URL?
    
    init() {
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = 15
        configuration.waitsForConnectivity = true
        session = URLSession(configuration: configuration)
    }
    
    func setEventHandler(_ handler: EventHandler?) {
        eventHandler = handler
    }
    
    static func normalizedServerURL(from address: String) throws -> URL {
        var value = address.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !value.isEmpty else { throw MAClientError.invalidServerAddress }
        if !value.contains("://") {
            value = "http://\(value)"
        }
        guard var components = URLComponents(string: value),
              let scheme = components.scheme?.lowercased(),
              scheme == "http" || scheme == "https",
              components.host != nil
        else {
            throw MAClientError.invalidServerAddress
        }
        components.path = components.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard let url = components.url else { throw MAClientError.invalidServerAddress }
        return url
    }
    
    static func fetchServerInfo(at serverURL: URL) async throws -> MAServerInfo {
        let url = serverURL.appending(path: "info")
        let (data, response) = try await URLSession.shared.data(from: url)
        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
            throw MAClientError.invalidResponse
        }
        return try JSONDecoder().decode(MAServerInfo.self, from: data)
    }
    
    static func login(
        at serverURL: URL,
        username: String,
        password: String
    ) async throws -> MALoginResponse {
        let requests = try loginRequests(
            for: serverURL.appending(path: "auth/login"),
            username: username,
            password: password
        )
        var firstError: Error?
        
        for request in requests {
            do {
                return try await performLoginRequest(request)
            } catch {
                firstError = firstError ?? error
            }
        }
        
        throw firstError ?? MAClientError.authenticationFailed("Login failed.")
    }
    
    static func loginRequests(
        for url: URL,
        username: String,
        password: String
    ) throws -> [URLRequest] {
        var currentJSONRequest = loginRequest(url: url, contentType: "application/json")
        currentJSONRequest.httpBody = try JSONSerialization.data(withJSONObject: [
            "provider_id": "builtin",
            "device_name": "MeloRemote",
            "credentials": [
                "username": username,
                "password": password
            ]
        ])
        
        var legacyJSONRequest = loginRequest(url: url, contentType: "application/json")
        legacyJSONRequest.httpBody = try JSONSerialization.data(withJSONObject: [
            "provider_id": "builtin",
            "device_name": "MeloRemote",
            "username": username,
            "password": password
        ])
        
        var formRequest = loginRequest(
            url: url,
            contentType: "application/x-www-form-urlencoded"
        )
        var form = URLComponents()
        form.queryItems = [
            URLQueryItem(name: "username", value: username),
            URLQueryItem(name: "password", value: password)
        ]
        formRequest.httpBody = Data((form.percentEncodedQuery ?? "").utf8)
        
        return [currentJSONRequest, legacyJSONRequest, formRequest]
    }
    
    private static func loginRequest(url: URL, contentType: String) -> URLRequest {
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 15
        request.setValue(contentType, forHTTPHeaderField: "Content-Type")
        return request
    }
    
    private static func performLoginRequest(_ request: URLRequest) async throws -> MALoginResponse {
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw MAClientError.invalidResponse
        }
        
        let authError = decodedAuthError(from: data)
        if http.statusCode == 401 {
            throw MAClientError.authenticationFailed(
                authError ?? "The username or password is incorrect."
            )
        }
        guard http.statusCode == 200 else {
            throw MAClientError.authenticationFailed(
                authError ?? "Login failed with status \(http.statusCode)."
            )
        }
        
        let result = try JSONDecoder().decode(MALoginResponse.self, from: data)
        guard result.resolvedAccessToken != nil else {
            throw MAClientError.authenticationFailed(result.error ?? "Music Assistant did not return a token.")
        }
        return result
    }
    
    private static func decodedAuthError(from data: Data) -> String? {
        guard let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }
        return object["error"] as? String
        ?? object["message"] as? String
        ?? object["detail"] as? String
    }
    
    func connect(serverURL: URL, token: String) async throws -> MAServerInfo {
        disconnect()
        baseURL = serverURL
        
        guard var components = URLComponents(url: serverURL, resolvingAgainstBaseURL: false) else {
            throw MAClientError.invalidServerAddress
        }
        components.scheme = components.scheme == "https" ? "wss" : "ws"
        components.path = "/ws"
        guard let webSocketURL = components.url else {
            throw MAClientError.invalidServerAddress
        }
        
        let socket = session.webSocketTask(with: webSocketURL)
        webSocket = socket
        socket.resume()
        
        let firstMessage = try await receiveValue(from: socket)
        let serverInfo = try firstMessage.decoded(as: MAServerInfo.self)
        guard serverInfo.schemaVersion >= serverInfo.minSupportedSchemaVersion else {
            throw MAClientError.incompatibleServer
        }
        
        listenerTask = Task { [weak self] in
            await self?.listen(on: socket)
        }
        pingTask = Task { [weak self] in
            await self?.keepAlive(on: socket)
        }
        
        _ = try await sendCommand(
            "auth",
            args: [
                "token": .string(token),
                "device_name": .string("MeloRemote")
            ]
        )
        return serverInfo
    }
    
    func disconnect() {
        listenerTask?.cancel()
        pingTask?.cancel()
        listenerTask = nil
        pingTask = nil
        webSocket?.cancel(with: .goingAway, reason: nil)
        webSocket = nil
        baseURL = nil
        failAllPending(with: MAClientError.disconnected)
    }
    
    func sendCommand<T: Decodable>(
        _ command: String,
        args: [String: JSONValue] = [:],
        as type: T.Type
    ) async throws -> T {
        let value = try await sendCommand(command, args: args)
        return try value.decoded(as: type)
    }
    
    func sendCommand(
        _ command: String,
        args: [String: JSONValue] = [:]
    ) async throws -> JSONValue {
        guard webSocket != nil else { throw MAClientError.disconnected }
        let messageID = UUID().uuidString.replacingOccurrences(of: "-", with: "")
        let payload: JSONValue = .object([
            "message_id": .string(messageID),
            "command": .string(command),
            "args": .object(args)
        ])
        let data = try JSONEncoder().encode(payload)
        guard let text = String(data: data, encoding: .utf8) else {
            throw MAClientError.invalidResponse
        }
        
        return try await withCheckedThrowingContinuation { continuation in
            pending[messageID] = continuation
            
            timeoutTasks[messageID] = Task { [weak self] in
                do {
                    try await Task.sleep(for: .seconds(20))
                } catch {
                    return
                }
                
                guard !Task.isCancelled else { return }
                
                await self?.resolvePending(
                    messageID,
                    with: .failure(
                        MAClientError.api(
                            "Music Assistant did not respond in time."
                        )
                    )
                )
            }
            
            Task { [weak self] in
                await self?.transmit(text, messageID: messageID)
            }
        }
    }
    
    private func transmit(_ text: String, messageID: String) async {
        guard let webSocket else {
            resolvePending(messageID, with: .failure(MAClientError.disconnected))
            return
        }
        do {
            try await webSocket.send(.string(text))
        } catch {
            resolvePending(messageID, with: .failure(error))
        }
    }
    
    private func listen(on socket: URLSessionWebSocketTask) async {
        do {
            while !Task.isCancelled {
                let value = try await receiveValue(from: socket)
                handle(value)
            }
        } catch {
            guard !Task.isCancelled else { return }
            failAllPending(with: error)
            eventHandler?(MAEventEnvelope(event: "disconnected", objectID: nil, data: nil))
        }
    }
    
    private func keepAlive(on socket: URLSessionWebSocketTask) async {
        while !Task.isCancelled {
            try? await Task.sleep(for: .seconds(30))
            guard !Task.isCancelled else { return }
            do {
                try await withCheckedThrowingContinuation {
                    (continuation: CheckedContinuation<Void, Error>) in
                    socket.sendPing { error in
                        if let error {
                            continuation.resume(throwing: error)
                        } else {
                            continuation.resume()
                        }
                    }
                }
            } catch {
                return
            }
        }
    }
    
    private func receiveValue(from socket: URLSessionWebSocketTask) async throws -> JSONValue {
        let message = try await socket.receive()
        let data: Data
        switch message {
        case .string(let text):
            data = Data(text.utf8)
        case .data(let messageData):
            data = messageData
        @unknown default:
            throw MAClientError.invalidResponse
        }
        return try JSONDecoder().decode(JSONValue.self, from: data)
    }
    
    private func handle(_ value: JSONValue) {
        guard let object = value.objectValue else { return }
        
        if let messageID = object["message_id"]?.stringValue {
            if let code = object["error_code"]?.doubleValue {
                let details = object["details"]?.stringValue ?? "Music Assistant API error \(Int(code))."
                resolvePending(messageID, with: .failure(MAClientError.api(details)))
                return
            }
            
            let isPartial = object["partial"]?.boolValue == true
            let result = object["result"] ?? .null
            if isPartial {
                if let values = result.arrayValue {
                    partialResults[messageID, default: []].append(contentsOf: values)
                }
                return
            }
            
            if let accumulated = partialResults.removeValue(forKey: messageID) {
                let finalValues = accumulated + (result.arrayValue ?? [])
                resolvePending(messageID, with: .success(.array(finalValues)))
            } else {
                resolvePending(messageID, with: .success(result))
            }
            return
        }
        
        if let event = object["event"]?.stringValue {
            eventHandler?(
                MAEventEnvelope(
                    event: event,
                    objectID: object["object_id"]?.stringValue,
                    data: object["data"]
                )
            )
        }
    }
    
    private func resolvePending(
        _ messageID: String,
        with result: Result<JSONValue, Error>
    ) {
        guard let continuation =
                pending.removeValue(forKey: messageID)
        else {
            return
        }
        
        timeoutTasks.removeValue(forKey: messageID)?.cancel()
        partialResults.removeValue(forKey: messageID)
        switch result {
        case .success(let value):
            continuation.resume(returning: value)
        case .failure(let error):
            continuation.resume(throwing: error)
        }
    }
    
    private func failAllPending(with error: Error) {
        let continuations = pending.values
        
        timeoutTasks.values.forEach {
            $0.cancel()
        }
        
        timeoutTasks.removeAll()
        pending.removeAll()
        partialResults.removeAll()
        
        continuations.forEach {
            $0.resume(throwing: error)
        }
    }
}

File: Android/app/src/main/java/com/versarepair/meloremote/data/MusicAssistantClient.kt
package com.versarepair.meloremote.data

import com.google.gson.Gson
import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.versarepair.meloremote.data.model.MAClientException
import com.versarepair.meloremote.data.model.MAEventEnvelope
import com.versarepair.meloremote.data.model.MALoginResponse
import com.versarepair.meloremote.data.model.MAServerInfo
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout
import okhttp3.Call
import okhttp3.Callback
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import java.io.IOException
import java.lang.reflect.Type
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

class MusicAssistantClient(
    private val gson: Gson = Gson(),
    private val httpClient: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .writeTimeout(15, TimeUnit.SECONDS)
        .pingInterval(30, TimeUnit.SECONDS)
        .build(),
) {
    private val pending = ConcurrentHashMap<String, CompletableDeferred<JsonElement>>()
    private val partialResults = ConcurrentHashMap<String, MutableList<JsonElement>>()

    @Volatile private var webSocket: WebSocket? = null
    @Volatile private var handshake: CompletableDeferred<MAServerInfo>? = null
    @Volatile private var closingIntentionally = false
    @Volatile private var eventHandler: ((MAEventEnvelope) -> Unit)? = null

    var baseUrl: HttpUrl? = null
        private set

    fun setEventHandler(handler: ((MAEventEnvelope) -> Unit)?) {
        eventHandler = handler
    }

    suspend fun fetchServerInfo(serverUrl: HttpUrl): MAServerInfo = withContext(Dispatchers.IO) {
        val request = Request.Builder().url(serverUrl.newBuilder().addPathSegment("info").build()).build()
        val response = httpClient.newCall(request).await()
        response.use {
            if (it.code != 200) throw MAClientException.InvalidResponse
            gson.fromJson(it.body.string(), MAServerInfo::class.java)
                ?: throw MAClientException.InvalidResponse
        }
    }

    suspend fun login(serverUrl: HttpUrl, username: String, password: String): MALoginResponse {
        val url = serverUrl.newBuilder().addPathSegments("auth/login").build()
        var firstError: Throwable? = null
        for (request in loginRequests(url, username, password)) {
            try {
                return performLoginRequest(request)
            } catch (error: Throwable) {
                if (firstError == null) firstError = error
            }
        }
        throw firstError ?: MAClientException.AuthenticationFailed("Login failed.")
    }

    fun loginRequests(url: HttpUrl, username: String, password: String): List<Request> {
        val currentPayload = JsonObject().apply {
            addProperty("provider_id", "builtin")
            addProperty("device_name", "MeloRemote")
            add("credentials", JsonObject().apply {
                addProperty("username", username)
                addProperty("password", password)
            })
        }
        val legacyPayload = JsonObject().apply {
            addProperty("provider_id", "builtin")
            addProperty("device_name", "MeloRemote")
            addProperty("username", username)
            addProperty("password", password)
        }
        val formPayload = "username=${username.formEncoded()}&password=${password.formEncoded()}"

        return listOf(
            requestWithBody(url, gson.toJson(currentPayload), "application/json"),
            requestWithBody(url, gson.toJson(legacyPayload), "application/json"),
            requestWithBody(url, formPayload, "application/x-www-form-urlencoded"),
        )
    }

    suspend fun connect(serverUrl: HttpUrl, token: String): MAServerInfo {
        disconnect()
        closingIntentionally = false
        baseUrl = serverUrl

        val socketUrl = serverUrl.newBuilder()
            .scheme(if (serverUrl.isHttps) "wss" else "ws")
            .encodedPath("/ws")
            .query(null)
            .build()
        val currentHandshake = CompletableDeferred<MAServerInfo>()
        handshake = currentHandshake
        val request = Request.Builder().url(socketUrl).build()
        webSocket = httpClient.newWebSocket(request, SocketListener())

        val info = try {
            withTimeout(15_000) { currentHandshake.await() }
        } catch (error: Throwable) {
            disconnect()
            throw error
        } finally {
            handshake = null
        }
        if (info.schemaVersion < info.minSupportedSchemaVersion) {
            disconnect()
            throw MAClientException.IncompatibleServer
        }

        sendCommand(
            "auth",
            mapOf(
                "token" to gson.toJsonTree(token),
                "device_name" to gson.toJsonTree("MeloRemote"),
            ),
        )
        return info
    }

    fun disconnect() {
        closingIntentionally = true
        handshake?.cancel()
        handshake = null
        webSocket?.close(1001, "Client disconnect")
        webSocket = null
        baseUrl = null
        failAllPending(MAClientException.Disconnected)
    }

    suspend fun sendCommand(
        command: String,
        args: Map<String, JsonElement> = emptyMap(),
    ): JsonElement {
        val socket = webSocket ?: throw MAClientException.Disconnected
        val messageId = UUID.randomUUID().toString().replace("-", "")
        val payload = JsonObject().apply {
            addProperty("message_id", messageId)
            addProperty("command", command)
            add("args", JsonObject().also { target -> args.forEach(target::add) })
        }
        val deferred = CompletableDeferred<JsonElement>()
        pending[messageId] = deferred
        if (!socket.send(gson.toJson(payload))) {
            pending.remove(messageId)
            throw MAClientException.Disconnected
        }
        return try {
            withTimeout(20_000) { deferred.await() }
        } catch (error: kotlinx.coroutines.TimeoutCancellationException) {
            throw MAClientException.Api("Music Assistant did not respond in time.")
        } finally {
            pending.remove(messageId)
            partialResults.remove(messageId)
        }
    }

    suspend fun <T> sendCommand(
        command: String,
        args: Map<String, JsonElement> = emptyMap(),
        type: Type,
    ): T = gson.fromJson(sendCommand(command, args), type)

    private suspend fun performLoginRequest(request: Request): MALoginResponse {
        val response = httpClient.newCall(request).await()
        response.use {
            val body = it.body.string()
            val authError = decodedAuthError(body)
            if (it.code == 401) {
                throw MAClientException.AuthenticationFailed(
                    authError ?: "The username or password is incorrect.",
                )
            }
            if (it.code != 200) {
                throw MAClientException.AuthenticationFailed(
                    authError ?: "Login failed with status ${it.code}.",
                )
            }
            val result = gson.fromJson(body, MALoginResponse::class.java)
                ?: throw MAClientException.InvalidResponse
            if (result.resolvedAccessToken.isNullOrBlank()) {
                throw MAClientException.AuthenticationFailed(
                    result.error ?: "Music Assistant did not return a token.",
                )
            }
            return result
        }
    }

    private fun handleMessage(text: String) {
        val value = runCatching { JsonParser.parseString(text) }.getOrNull() ?: return
        if (!value.isJsonObject) return
        val objectValue = value.asJsonObject

        val activeHandshake = handshake
        if (activeHandshake != null && !activeHandshake.isCompleted && objectValue.has("server_id")) {
            runCatching { gson.fromJson(objectValue, MAServerInfo::class.java) }
                .onSuccess(activeHandshake::complete)
                .onFailure(activeHandshake::completeExceptionally)
            return
        }

        val messageId = objectValue.stringOrNull("message_id")
        if (messageId != null) {
            if (objectValue.has("error_code")) {
                val code = objectValue["error_code"]?.asInt ?: -1
                val details = objectValue.stringOrNull("details")
                    ?: "Music Assistant API error $code."
                pending.remove(messageId)?.completeExceptionally(MAClientException.Api(details))
                partialResults.remove(messageId)
                return
            }
            val result = objectValue["result"] ?: com.google.gson.JsonNull.INSTANCE
            if (objectValue["partial"]?.asBoolean == true) {
                if (result.isJsonArray) {
                    partialResults.computeIfAbsent(messageId) { mutableListOf() }
                        .addAll(result.asJsonArray)
                }
                return
            }
            val accumulated = partialResults.remove(messageId)
            val resolved = if (accumulated != null) {
                JsonArray().apply {
                    accumulated.forEach(::add)
                    if (result.isJsonArray) result.asJsonArray.forEach(::add)
                }
            } else {
                result
            }
            pending.remove(messageId)?.complete(resolved)
            return
        }

        objectValue.stringOrNull("event")?.let { event ->
            eventHandler?.invoke(
                MAEventEnvelope(
                    event = event,
                    objectId = objectValue.stringOrNull("object_id"),
                    data = objectValue["data"],
                ),
            )
        }
    }

    private fun failAllPending(error: Throwable) {
        pending.values.forEach { it.completeExceptionally(error) }
        pending.clear()
        partialResults.clear()
    }

    private inner class SocketListener : WebSocketListener() {
        override fun onMessage(webSocket: WebSocket, text: String) = handleMessage(text)

        override fun onMessage(webSocket: WebSocket, bytes: ByteString) = handleMessage(bytes.utf8())

        override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
            handshake?.completeExceptionally(t)
            failAllPending(t)
            if (!closingIntentionally) {
                eventHandler?.invoke(MAEventEnvelope("disconnected", null, null))
            }
        }

        override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
            failAllPending(MAClientException.Disconnected)
            if (!closingIntentionally) {
                eventHandler?.invoke(MAEventEnvelope("disconnected", null, null))
            }
        }
    }

    companion object {
        private val JSON = "application/json; charset=utf-8".toMediaType()

        fun normalizedServerUrl(address: String): HttpUrl {
            var value = address.trim()
            if (value.isEmpty()) throw MAClientException.InvalidServerAddress
            if (!value.contains("://")) value = "http://$value"
            val url = value.toHttpUrlOrNull() ?: throw MAClientException.InvalidServerAddress
            if (url.scheme != "http" && url.scheme != "https") {
                throw MAClientException.InvalidServerAddress
            }
            return url.newBuilder()
                .encodedPath(url.encodedPath.trimEnd('/').ifEmpty { "/" })
                .build()
        }

        private fun requestWithBody(url: HttpUrl, body: String, contentType: String): Request =
            Request.Builder()
                .url(url)
                .post(body.toRequestBody(contentType.toMediaType()))
                .build()

        private fun decodedAuthError(body: String): String? {
            val value = runCatching { JsonParser.parseString(body).asJsonObject }.getOrNull() ?: return null
            return value.stringOrNull("error")
                ?: value.stringOrNull("message")
                ?: value.stringOrNull("detail")
        }

        private fun String.formEncoded(): String =
            URLEncoder.encode(this, StandardCharsets.UTF_8.toString())
    }
}

private fun JsonObject.stringOrNull(key: String): String? =
    get(key)?.takeUnless { it.isJsonNull }?.runCatching { asString }?.getOrNull()

private suspend fun Call.await(): Response = suspendCancellableCoroutine { continuation ->
    continuation.invokeOnCancellation { cancel() }
    enqueue(object : Callback {
        override fun onFailure(call: Call, e: IOException) {
            if (continuation.isActive) continuation.resumeWithException(e)
        }

        override fun onResponse(call: Call, response: Response) {
            if (continuation.isActive) continuation.resume(response) else response.close()
        }
    })
}
